### ISSUE TITLE
Variable can be immutable intention

### ISSUE THREAD

**Issue by mkaput** [intellij-rust/intellij-rust#1004]
Add intention to detect effectively immutable variables, like:

```rust
struct St(i32);

fn foo(v: &St) {
    unimplemented!();
}

fn foo_mut(v: &mut St) {
    unimplemented!();
}

trait Bar {
    fn bar(&self);
    fn bar_mut(&mut self);
}

impl Bar for St {
    fn bar(&self) {
        unimplemented!();
    }
    
    fn bar_mut(&mut self) {
        unimplemented!();
    }
}

fn main() {
    // This variable is annotated because we do not
    // use its mutability (it's effectively immutable)
    let mut triggers = St(10);
    foo(&triggers);
    triggers.bar();

    // This variable isn't annotated because of one of these calls
    let mut does_not_trigger= St(10);
    foo_mut(&mut does_not_trigger);
    does_not_trigger.bar_mut();
    does_not_trigger= St(20);

    // What about macros??
    // Maybe simple token scanning for matching identifier
    // would work (as far as I remember macro hygiene prohibits
    // implicit variable usages). Or we can just ignore them.
    some_funny_macro!(does_not_trigger);
}
```

**Comment by farodin91** [comment]
I would like to work on it.

**Comment by Kobzol** [comment]
@ortem should this be an additional cached analysis pass, similar to `RsInferenceContextOwner.liveness`? It would scan local usages of bindings in variables, parameters, match bindings, etc., check if they are used within any context that requires mutability and if not, it would mark them. If you agree with this direction, I'd try to implement it.

**Comment by ortem** [comment]
@Kobzol Yes, I think it can be an additional cached analysis pass like liveness. I don't mind if you try to implement it 👍

**Comment by Kobzol** [comment]
Hmm, I found out that there is actually already an inspection for this - `RsVariableMutableInspection`, but it's disabled by default and I didn't know about it, aargh! :D I think that it's a shame that such a useful inspection is disabled. I'll transform it to a lint inspection and enable it by default to gather possible false positives.

**Comment by ortem** [comment]
@Kobzol `RsVariableMutableInspection` [had been created](https://github.com/intellij-rust/intellij-rust/pull/1143) before our control-flow analysis was implemented, so this inspection is very simple and doesn't support lots of real-world cases. So it has many false-positives and then should not be enabled by default.

The proper implementation should use our data-flow analysis framework like liveness and move analysis (a part of borrow checker implemented without MIR) do. Probably it should even be a part of our borrow checker because they are very similar. As far as I remember, this analysis is a part of borrow checker in rustc implementation. So you can try to implement it as a new analysis pass or as an extension of our borrow (basically move) checker.

**Comment by Kobzol** [comment]
Okay, thanks for the explanation :) Do you think that it is worth it to enable the current inspection in nightly in the meantime to gather possible false positives or not?

Also, it seems that the current resolve machinery is actually quite good and it can follow basic control-flow, i.e.
```rust
let mut f = 5;
f *= 6;
let mut f = 6; // warning is shown only here
```
works. So far I only saw false negatives, do you have an example of a false positive that happens with the current inspection? It would help me to understand how data-flow analysis would help. I ran the inspection on my mid-sized Rust project and also on `tokio` and I didn't find any false positives. (there were some false negatives).

**Comment by ortem** [comment]
@Kobzol You're right, the current implementation doesn't produce false-positives because it is very limited. I thought it is disabled because of lots of false-positives, but now I'm not sure why it is disabled 🤔

However, `RsVariableMutableInspection` is too limited and supports **very** few cases. It might not be noticeable while testing on real projects like `tokio` because real projects usually don't contain many warnings. But this inspection can hardly be improved significantly without using data-flow. Consider this example:
```Rust
fn test_write_once(flag: bool) {
    let mut x: i32; // should warn
    if flag {
        x = 1;
    } else {
        x = 2;
    }
}
````
Variable `x` should be immutable because there is only one write-access. But it is not obvious when traversing the program tree or even using `ReferencesSearch` without taking the control-flow into account.

**Comment by Kobzol** [comment]
Thank you, I didn't consider this case, now it makes sense. I think that the inspection is still useful in the current state, although it would probably have some false negatives. But I'm not sure if it's worth enabling it by default now since the plan is to rewrite it anyway. I'll try to come up with a data-flow based prototype.

**Comment by Kobzol** [comment]
@ortem I tried to clone and modify the liveness inspection, but since the code does not have many comments, I'm not sure that I understand the concepts right. Could you please give me a short description of  what `UsagePath` represents?

I more or less understand the dataflow concepts, `gen` will probably be used for mutating usages of each binding instead of all usages, but what I don't understand is how to solve the example code that you have posted using the current machinery. I thought that the implemented data-flow already supports it, but it seems that it is not the case:
```rust
fn main() {
    let a;
    if true {
        a = 6;
    } else {
        a = 5;
    }
}
```
Rustc gives unused variable warning for `a` (in addition to dead assignment, but that is not implemented yet in the plugin), but `RsLivenessInspection` does not annotate it, because it does not consider the (first) write operation for declarations without init to be an init operation. Maybe we should first modify the dataflow machinery to support this use-case and then use it for the unused mutability inspection?

**Comment by ortem** [comment]
@Kobzol Currently, our liveness analysis does not distinguish between **write** and **read** operations and treats all of them as just usages. All usages are considered as `gen` and only declarations are considered as `kill`. To support dead assignments and `variable is assigned but never used`, the liveness analysis should handle write and read operations differently in terms of `gen` and `kill`. I planned to implement it someday, but you can try doing it.

Unused mutability analysis should calculate the number of write operations for each binding. If there is only one write access to the mutable binding, it should be annotated as potentially immutable.

If you are interested in dataflow analysis, it can be very helpful to read some related literature or articles; I read a bit of "Data Flow Analysis: Theory and Practice" by Khedker, Sanyal, Karkare before implementing this analysis. I believe that it is not necessary but can help.

**Comment by Kobzol** [comment]
Thanks, I'll check that book. The current dataflow is almost good enough for this inspection, because mut is often required for calling `&mut self` methods or taking `&mut` references and not reassigning the variable and that can be sort of done with read usages. But of course the dataflow needs to be modified to handle writes too.
